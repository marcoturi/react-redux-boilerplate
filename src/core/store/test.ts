/**
 * This function is used to mock the state of the store to test the selectors.
 * However, is quite fragile, and it's better to use the real store.
 * https://github.com/reduxjs/redux-toolkit/discussions/4016
 */
export function getMockedState({
  functionName,
  parameters,
  data,
}: {
  functionName: string;
  parameters?: any;
  data?: any;
}) {
  const state: any = {};
  if (data) {
    state.data = data;
  }
  return {
    api: {
      queries: {
        [`${functionName}(${
          parameters ? JSON.stringify(parameters) : 'undefined'
        })`]: state,
      },
    },
  } as any;
}
