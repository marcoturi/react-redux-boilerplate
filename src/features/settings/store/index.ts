import { settingsSlice } from './settings.slice';
import { loadState } from './settings.thunk';

const SettingsActions = {
  ...settingsSlice.actions,
  loadState,
};

export { SettingsActions };

export * as SettingsSelectors from './settings.selector';
