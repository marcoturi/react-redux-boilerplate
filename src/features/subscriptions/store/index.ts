export * as SubscriptionApi from './subscription.api';
export * as SubscriptionSelectors from './subscription.selector';
