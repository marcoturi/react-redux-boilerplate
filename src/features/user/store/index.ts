export * as UserSelectors from './user.selector';
export * as UserApi from './user.api';
