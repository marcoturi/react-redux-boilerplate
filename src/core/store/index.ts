import env from '../config/env';
import baseApi from '@/core/store/api';
import { storageMiddleware } from '@/features/settings/store/settings.effect';
import { settingsSlice } from '@/features/settings/store/settings.slice';
import { combineReducers, configureStore } from '@reduxjs/toolkit';
import * as Sentry from '@sentry/react';

Sentry.init({
  dsn: env.SENTRY_DSN,
  release: env.VERSION,
  enabled: !env.IS_DEV,
  environment: env.NODE_ENV,
  tracesSampleRate: 0.1,
  attachStacktrace: true,
  integrations: [new Sentry.Replay()],
  replaysSessionSampleRate: 0.1,
  replaysOnErrorSampleRate: 1,
});

const sentryReduxEnhancer = Sentry.createReduxEnhancer();

export const rootReducer = combineReducers({
  api: baseApi.reducer,
  settings: settingsSlice.reducer,
});

export const storeConfig = {
  reducer: rootReducer,
  devTools: env.IS_DEV,
  enhancers: (getDefaultEnhancers) =>
    // eslint-disable-next-line unicorn/prefer-spread
    getDefaultEnhancers().concat(sentryReduxEnhancer),
  middleware: (getDefaultMiddleware) =>
    // eslint-disable-next-line unicorn/prefer-spread
    getDefaultMiddleware().concat(
      baseApi.middleware,
      storageMiddleware.middleware,
    ) as any,
};

export const store = configureStore(storeConfig);
