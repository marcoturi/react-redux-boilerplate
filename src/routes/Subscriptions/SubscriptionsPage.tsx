import { Container } from '@/UI/Elements/Container';
import { Text } from '@/UI/Elements/Text';
import { SubscriptionList } from '@/features/subscriptions/components';
import { SubscriptionFilter } from '@/features/subscriptions/components/SubscriptionFilter';

function SubscriptionsPage() {
  return (
    <Container>
      <Text
        size="7"
        as="p"
      >
        Subscriptions
      </Text>
      <div className="flex items-center space-x-2 pt-5">
        <SubscriptionFilter />
      </div>
      <SubscriptionList />
    </Container>
  );
}

export default SubscriptionsPage;
