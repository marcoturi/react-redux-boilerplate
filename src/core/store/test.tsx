import {  RootState } from '@/core/store/types';
import { render } from '@testing-library/react';
import type { RenderOptions } from '@testing-library/react';
import React, { PropsWithChildren } from 'react';
import { Provider } from 'react-redux';
import { configureStore } from '@reduxjs/toolkit';
import { storeConfig } from '@/core/store/index';

const setupStore = (preloadedState?: Partial<RootState>) => configureStore({
  ...storeConfig,
  preloadedState,
});

type AppStore = ReturnType<typeof setupStore>;

// This type interface extends the default options for render from RTL, as well
// as allows the user to specify other things such as initialState, store.
interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
  preloadedState?: Partial<RootState>;
  store?: AppStore;
}

export function renderWithProviders(
  ui: React.ReactElement,
  {
    preloadedState = {},
    // Automatically create a store instance if no store was passed in
    store = setupStore(preloadedState),
    ...renderOptions
  }: ExtendedRenderOptions = {},
) {
  // eslint-disable-next-line @typescript-eslint/ban-types
  function Wrapper({ children }: PropsWithChildren<{}>): JSX.Element {
    return <Provider store={store}>{children}</Provider>;
  }

  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
}

// This type interface extends the default options for render from RTL, as well
// as allows the user to specify other things such as initialState, store.
interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
  preloadedState?: Partial<RootState>;
  store?: AppStore;
}

/**
 * This function is used to mock the state of the store to test the selectors.
 * However, is quite fragile, and it's better to use the real store.
 * https://github.com/reduxjs/redux-toolkit/discussions/4016
 */
export function getMockedState({
  functionName,
  parameters,
  data,
}: {
  functionName: string;
  parameters?: any;
  data?: any;
}) {
  const state: any = {};
  if (data) {
    state.data = data;
  }
  return {
    api: {
      queries: {
        [`${functionName}(${
          parameters ? JSON.stringify(parameters) : 'undefined'
        })`]: state,
      },
    },
  } as any;
}
