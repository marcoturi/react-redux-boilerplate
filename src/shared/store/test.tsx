import { storeConfig } from '@/shared/store/index';
import { RootState } from '@/shared/store/types';
import { configureStore } from '@reduxjs/toolkit';
import { render } from 'vitest-browser-react';
import React from 'react';
import { Provider } from 'react-redux';

export const setupStore = (preloadedState?: Partial<RootState>) =>
  configureStore({
    ...storeConfig,
    preloadedState,
  });

type AppStore = ReturnType<typeof setupStore>;

// This type interface extends the default options for render from RTL, as well
// as allows the user to specify other things such as initialState, store.
interface ExtendedRenderOptions extends Omit<any, 'queries'> {
  preloadedState?: Partial<RootState>;
  store?: AppStore;
}

export function renderWithProviders(
  ui: React.ReactElement,
  {
    preloadedState = {},
    // Automatically create a store instance if no store was passed in
    store = setupStore(preloadedState),
  }: ExtendedRenderOptions = {},
) {
  const rend = render(<Provider store={store}>{ui}</Provider>);
  return rend;
}

// This type interface extends the default options for render from RTL, as well
// as allows the user to specify other things such as initialState, store.
interface ExtendedRenderOptions extends Omit<any, 'queries'> {
  preloadedState?: Partial<RootState>;
  store?: AppStore;
}
