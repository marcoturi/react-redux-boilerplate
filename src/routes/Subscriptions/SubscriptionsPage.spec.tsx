import SubscriptionsPage from '@/routes/Subscriptions/SubscriptionsPage';
import { renderWithProviders } from '@/shared/store/test';
// import { userEvent } from '@vitest/browser/context';
import { page } from '@vitest/browser/context'
import { expect, it, describe, vi } from 'vitest';

describe('SubscriptionsPage', () => {
  // it('Should load page title', async () => {
  //   const screen = renderWithProviders(<SubscriptionsPage />);
  //   await expect.element(page.getByText(/subscriptions/i)).toBeVisible()
  //   expect(screen.container).toMatchSnapshot()
  // });
  it('Should load subscriptions', async () => {
     renderWithProviders(<SubscriptionsPage />);
    vi.waitFor(async () => {
      const a = page.getByTestId('subscription-item').elements();
      console.log(a);
      await expect.element(page.getByTestId('subscription-item')).toBeInTheDocument()
    })

    // expect(screen.locator('[data-testid="subscription-item"').length).toBe(3);
  });

  // it('Should handle a change in filters', async () => {
  //   const screen = renderWithProviders(<SubscriptionsPage />);
  //   const user = userEvent.setup();
  //
  //   // Filter should be present
  //   const optionSelect = screen.getByRole('combobox');
  //   expect(optionSelect).toBeInTheDocument();
  //   expect(screen.queryByText('All')).toBeInTheDocument();
  //   expect(screen.queryByText('One Time Purchase')).not.toBeInTheDocument();
  //
  //   // Options should be clickable
  //   await user.click(optionSelect);
  //   const oneTimePurchase = screen.getByRole('option', {
  //     name: 'One Time Purchase',
  //   });
  //   expect(oneTimePurchase).toBeInTheDocument();
  //   await user.click(oneTimePurchase);
  //   expect(screen.queryByText('One Time Purchase')).toBeInTheDocument();
  //   expect(screen.queryByText('All')).not.toBeInTheDocument();
  // });
});
