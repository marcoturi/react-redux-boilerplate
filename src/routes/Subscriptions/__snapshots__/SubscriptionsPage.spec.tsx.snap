// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SubscriptionsPage > Should load page title 1`] = `
<div>
  <div
    class="rt-Container rt-r-size-4 w-11/12 pt-10"
  >
    <div
      class="rt-ContainerInner"
    >
      <p
        class="rt-Text rt-r-size-7 leading-7 [&:not(:first-child)]:mt-6"
      >
        Subscriptions
      </p>
      <div
        class="flex items-center space-x-2 pt-5"
      >
        <button
          aria-autocomplete="none"
          aria-controls="radix-:r0:"
          aria-expanded="false"
          class="flex h-10 items-center justify-between rounded-md border border-input bg-background px-3 py-2 text-sm ring-offset-background placeholder:text-muted-foreground focus:outline-none focus:ring-2 focus:ring-ring focus:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50 [&>span]:line-clamp-1 w-[180px]"
          data-state="closed"
          data-testid="filter-trigger"
          dir="ltr"
          role="combobox"
          type="button"
        >
          <span
            style="pointer-events: none;"
          >
            All
          </span>
          <svg
            aria-hidden="true"
            class="lucide lucide-chevron-down h-4 w-4 opacity-50"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="m6 9 6 6 6-6"
            />
          </svg>
        </button>
      </div>
      <div
        class="py-8"
      >
        <div
          class="flex items-center space-x-4 py-5"
        >
          <div
            class="animate-pulse bg-muted h-12 w-12 rounded-full"
          />
          <div
            class="space-y-2"
          >
            <div
              class="animate-pulse rounded-md bg-muted h-4 w-[250px]"
            />
            <div
              class="animate-pulse rounded-md bg-muted h-4 w-[200px]"
            />
          </div>
        </div>
        <div
          class="flex items-center space-x-4 py-5"
        >
          <div
            class="animate-pulse bg-muted h-12 w-12 rounded-full"
          />
          <div
            class="space-y-2"
          >
            <div
              class="animate-pulse rounded-md bg-muted h-4 w-[250px]"
            />
            <div
              class="animate-pulse rounded-md bg-muted h-4 w-[200px]"
            />
          </div>
        </div>
        <div
          class="flex items-center space-x-4 py-5"
        >
          <div
            class="animate-pulse bg-muted h-12 w-12 rounded-full"
          />
          <div
            class="space-y-2"
          >
            <div
              class="animate-pulse rounded-md bg-muted h-4 w-[250px]"
            />
            <div
              class="animate-pulse rounded-md bg-muted h-4 w-[200px]"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
